slurm/install/installlib.py

```diff
@@ -427,3 +493,183 @@ def set_hostname(
             "import jetpack.converge as jc; jc._send_installation_status('warning')",
         ],
     )
+
+
+class CCNode:
+    """
+    Simple CycleCloud node representation.
+    """
+    def __init__(
+        self,
+        name: str,
+        nodearray_name: str,
+        hostname: str,
+        private_ipv4: str,
+        status: str,
+        software_configuration: Dict,
+    ) -> None:
+        self.name = name
+        self.nodearray_name = nodearray_name
+        self.hostname = hostname
+        self.private_ipv4 = private_ipv4
+        self.status = status
+        self.software_configuration = software_configuration
+
+    def to_dict(self) -> Dict:
+        ret = {}
+        for attr in dir(self):
+            if attr.startswith("_"):
+                continue
+            if attr == "to_dict":
+                continue
+            val = getattr(self, attr)
+            if hasattr(val, "__call__"):
+                continue
+            ret[attr] = val
+        return ret
+
+    def is_failed(self) -> bool:
+        return self.status == "Failed"
+
+    def is_ready(self) -> bool:
+        return self.status == "Ready"
+
+    def is_booting(self) -> bool:
+        return self.status != "Ready"
+
+    def __repr__(self) -> str:
+        return f"CCNode({self.to_dict()})"
+
+    def __str__(self) -> str:
+        return repr(self)
+
+    def __eq__(self, other: object) -> bool:
+        if hasattr(other, "to_dict"):
+            return self.to_dict() == getattr(other, "to_dict")()
+        return False
+
+
+def cluster_status(config: Dict) -> Dict:
```

Comment:
I am a bit concerned about making REST calls here in installlib. Purely because we have been using this lib as a way to install slurm. Is "scalelib" which is already configured to make calls to CC a better place for this? I may be wrong about this, but to me it looks like its a better fit there.

Reply:
As we discussed offline, I do understand that concern but scalelib requires quite a few dependencies and was not itself meant to be used just as an installer. Requiring so many packages to be installed on execute nodes in a CX environment just to install slurm (but not azslurm which uses scalelib) is just not a tradeoff we can make at this time.

Reply:
I dont think the intent here is to use scalelib as an installer. there are a couple of issues i am trying to point out here:

First, it is that installlib should just be concerned with doing installation of rpms, general configuration, and enabling/disabling services. Based on the change we just merged, PR #165, Its not even concerned with starting stopping services. So currently the code in installlib is indempotent, and this change breaks that.

my concern is more with blocking on different nodes in the topology to come up, and second is the obvious code duplication we are having to do here because we cannot include scalelib.

everything we are doing in installlib, can perfectly happen without waiting on the ha node to come up. And based on that we dont really need to have this code at all.

What we can do instead is-- in start_services.sh, just ping the HA node (based on the slurm config) to see if its pingable. and if it is then start slurmctld. Then any failure automatically corresponds to just failure of starting stopping services. Failure of HA node to come up is no different than failing to start slurmdbd (to give an example), and is a perfectly valid reason to fail the script. 

that change would make this PR much more cleaner and self contained, and avoid code duplication.

Reply:
I forgot to mention, I fully agree with your assessment that scalelib is not the right place, but what i am suggesting is that this code need not exist at all.